File: golang-devildyw/middle/L3无重复字符的最长子串.go
package main

func main() {

}

// 滑动窗口罢了
func lengthOfLongestSubstring(s string) int {
	mapp := make(map[byte]struct{})
	result := 0
	left := 0
	for i := 0; i < len(s); i++ {
		for {
			if _, ok := mapp[s[i]]; ok {
				delete(mapp, s[left])
				left++
			} else {
				break
			}
		}
		mapp[s[i]] = struct{}{}
		result = max(result, i-left+1)
	}

	return result
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L3无重复字符的最长子串.java
package top.dyw.leetcode.Hot100;

import java.util.HashMap;

public class L3无重复字符的最长子串 {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap();

        int result = 0;
        int left = 0;
        for (int right=0; right<s.length(); right++) {
            if (map.containsKey(s.charAt(right))) {
                left = Math.max(left, map.get(s.charAt(right))+1);
            }
            map.put(s.charAt(right), right);
            result = Math.max(result, right-left+1);
        }

        return result;
    }
}
